Add capability to compute average of a variable over the exposed portion of a boundary

bwspenc opened the pull request:
closes #23949

bwspenc committed:
Add ExposedSideAverageValue to compute average value over exposed portion of sideset ref #23949

bwspenc committed:
Minor cleanup items noticed when implementing #23949

Removed a shadow member variable, added doxygen comments

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1441721/) on 0ff3c79 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/23950/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1441756/) on 0ff3c79 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>40cceb</th>
      <th colspan=4>#23950 0ff3c7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.91%</td>
      <td>84.91%</td>
      <td>-0.00%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>84351</td>
      <td>84347</td>
      <td>-4</td>
      <td>1</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>14987</td>
      <td>14991</td>
      <td>+4</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23950/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23950/coverage/framework)


## Modules coverage

### Heat conduction

<table>
  <thead>
    <tr>
      <th></th>
      <th>40cceb</th>
      <th colspan=4>#23950 0ff3c7</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.11%</td>
      <td>89.14%</td>
      <td>+0.03%</td>
      <td>95.45%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>3558</td>
      <td>3579</td>
      <td>+21</td>
      <td>21</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>435</td>
      <td>436</td>
      <td>+1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/23950/coverage_diff/heat_conduction)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/23950/coverage/heat_conduction)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/23950/coverage/xfem)

</details>



This comment will be updated on new commits.

lindsayad commented on modules/heat_conduction/include/postprocessors/ExposedSideAverageValue.h:
```diff
@@ -0,0 +1,38 @@
+   * @return exposed face volume
+   */
+  virtual Real volume() override;
+  /// Reference to SelfShadowSideUserObject, which does the determination of which
```
```suggestion

  /// Reference to SelfShadowSideUserObject, which does the determination of which
```

lindsayad commented on modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:
```diff
@@ -0,0 +1,63 @@
+    _self_shadow(getUserObject<SelfShadowSideUserObject>("self_shadow_uo"))
+{
+  if (_self_shadow.useDisplacedMesh() != getParam<bool>("use_displaced_mesh"))
+    paramWarning("use_displaced_mesh",
```
Should this be an error?

bwspenc replied:
Yes -- changed it to an error.

lindsayad commented on modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:
```diff
@@ -0,0 +1,63 @@
+  Real curr_exposed_side_volume = 0.0;
+  const SelfShadowSideUserObject::SideIDType id(_current_elem->id(), _current_side);
+  const int illumination = _self_shadow.illumination(id);
+  for (unsigned int qp = 0; qp < _qrule->n_points(); qp++)
```
```suggestion
  for (const unsigned int qp : make_range(_qrule->n_points()))
```

lindsayad commented on modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:
```diff
@@ -0,0 +1,63 @@
+  for (unsigned int qp = 0; qp < _qrule->n_points(); qp++)
+  {
+    // tests if the bit at position _qp is set
+    if (illumination & (1 << qp))
```
https://wiki.sei.cmu.edu/confluence/display/c/INT13-C.+Use+bitwise+operators+only+on+unsigned+operands suggests bitwise operations should only ever be performed on unsigned integers

bwspenc replied:
Agreed. I changed `SelfShadowSideUserObject::illumination()` to return an unsigned int, and changed the variable types accordingly.

lindsayad reviewed (changes requested):
Seems like we should be updating some requirements for the tests you're modifying?

bwspenc committed:
Apply suggestions from code review

Co-authored-by: Alex Lindsay <alexlindsay239@gmail.com>

bwspenc committed:
Address review comments

This includes generating an error message for usage of SideAverageValue
on the centerline of an axisymmetric model.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1545179/) on 1b5f0ec wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23950/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23950/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 395ab6da47fe9a63aab358a9ca122e2fd6dd943d`

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1545227/) on 1a062f0 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23950/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23950/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 395ab6da47fe9a63aab358a9ca122e2fd6dd943d`

bwspenc commented on modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:
```diff
@@ -0,0 +1,63 @@
+{
+  InputParameters params = SideAverageValue::validParams();
+  params.addClassDescription("Computes the average value of a variable on the "
+                             "exposed portion of a sideset. Note that this cannot be used on the "
+                             "centerline of an axisymmetric model.");
```
This class derives from `SideAverageValue` (in `framework/src/postprocessors`), from which I carried over the same note in its class description. The actual integration over the sides is done in `SideIntegralPostprocessor::computeIngegral()`. I believe that if a qp is on the centerline, its value of `_coord` will be zero. If this happens, we'll contribute zero to the integral (which is the correct thing to do), and `SideAverageValue::volume()` will return zero. If we only have faces on the centerline, we would get a divide by zero in `SideAverageValue::getValue(), which I guess we should check against. Maybe we should talk for a few minutes about how to handle this.

bwspenc replied:
I added a check for the divide by zero mentioned above.

lindsayad reviewed (dismissed):
Looks good @bwspenc ! Merge when you're ready (if tests pass 😄 )

GiudGiud commented on modules/heat_conduction/doc/content/source/postprocessors/ExposedSideAverageValue.md:
```diff
@@ -0,0 +1,13 @@
+
+This is used to compute the average of a variable over only the portion of a specified surface that is exposed, using the [SelfShadowSideUserObject](SelfShadowSideUserObject.md) to determine which portions of the surface are exposed and which are shadowed. This computation is identical to that performed in [SideAverageUserObject](SideAverageUserObject.md), except that only portions of the surface ...
+
+!listing directional_flux_bc/2d.i block=Postprocessors/ave_v_exposed
```
This is not useful unless you add text to say what s being done.
The !syntax inputs /Postprocessors/ExposedSideAverageValue lists this input already

bwspenc replied:
Removed

GiudGiud commented on modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:
```diff
@@ -0,0 +1,63 @@
+#include "ExposedSideAverageValue.h"
+#include "SelfShadowSideUserObject.h"
+
+registerMooseObject("MooseApp", ExposedSideAverageValue);
```
Not the right app

bwspenc replied:
Good catch! I fixed that.

GiudGiud commented on modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:
```diff
@@ -0,0 +1,63 @@
+  const SelfShadowSideUserObject::SideIDType id(_current_elem->id(), _current_side);
+  const unsigned int illumination = _self_shadow.illumination(id);
+  // tests if the bit at position _qp is set
+  if (illumination & (1 << _qp))
```
this `1 << _qp` thing does not seem right. where did it come from? would you have a minute to explain it to me?

dschwen replied:
It is right. `<<` is the left shift operator. It generates the values 1 2 4 8 for _qp 0 1 2 3 (which are bit masks to test bit 0 1 2 3). The illumination state of all QPs of a side element is stored as a bitmask.

bwspenc replied:
I believe this is implemented correctly like @dschwen commented. However, I also think this is a confusing way to get that information from `SelfShadowSideUserObject`. I didn't really want to touch the implementation of `SelfShadowSideUserObject` in this PR. Maybe that could be improved in a follow-on PR?

GiudGiud commented on test/tests/userobjects/side_user_object_no_boundary_error/tests:
```diff
@@ -26,4 +26,13 @@
+    cli_args = "Postprocessors/avg/boundary='left' Problem/coord_type=RZ"
+    mesh_mode = replicated
+    expect_err = "The total area of the boundary is zero. This could be due to using a boundary on the centerline of an axisymmetric model."
+    requirement = "The system shall error if SideAverageValue is applied to a boundary on the centerline of an axisymmetric model."
```
```suggestion
    requirement = "The system shall error if the side average value of a variable is requested on a boundary that is the centerline of an axisymmetric model."
```

bwspenc replied:
Applied changes

GiudGiud commented on modules/heat_conduction/src/postprocessors/ExposedSideAverageValue.C:
```diff
@@ -0,0 +1,63 @@
+                             "centerline of an axisymmetric model.");
+  params.addRequiredParam<UserObjectName>(
+      "self_shadow_uo",
+      "SelfShadowSideUserObject that calculates the illumination state of a side set");
```
```suggestion
      "SelfShadowSideUserObject that calculates the illumination state of element sides in a side set");
```

bwspenc committed:
Address more review comments

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1546588/) on 5d07a6e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/23950/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/23950/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 21d9616abc39bb56b5fc5b6218b88cc7f9d6064a`

GiudGiud reviewed (dismissed):
bitwise approval

lindsayad commented on modules/heat_conduction/doc/content/source/postprocessors/ExposedSideAverageValue.md:
```diff
@@ -0,0 +1,11 @@
+
+!syntax description /Postprocessors/ExposedSideAverageValue
+
+This is used to compute the average of a variable over only the portion of a specified surface that is exposed, using the [SelfShadowSideUserObject](SelfShadowSideUserObject.md) to determine which portions of the surface are exposed and which are shadowed. This computation is identical to that performed in [SideAverageUserObject](SideAverageUserObject.md), except that only portions of the surface ...
```
```suggestion
This is used to compute the average of a variable over only the portion of a specified surface that is exposed, using the [SelfShadowSideUserObject](SelfShadowSideUserObject.md) to determine which portions of the surface are exposed and which are shadowed. This computation is identical to that performed in [SideAverageValue](SideAverageValue.md), except that only portions of the surface that are exposed (i.e., not shadowed) are included in the calculation.
```

bwspenc replied:
Fixed. Thanks for finding this issue!

bwspenc committed:
Fix documentation build issue

bwspenc committed:
Move coord_type from Problem to Mesh in definition of new test
